File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
  let leftBrackets = 0, rightBrackets = 0, max = 0;
  
  //Scan Left to Right 
  for (let i=0; i<s.length; i++) {
    if (s[i] == '(') {
      leftBrackets++;
    } else {
      rightBrackets++;
    }
    
    if (leftBrackets === rightBrackets) max = Math.max(max, 2*rightBrackets);
    
    if (rightBrackets > leftBrackets) {
      leftBrackets = 0;
      rightBrackets = 0;
    }
  }
  
  //Scan Right to Left
  leftBrackets = 0;
  rightBrackets = 0;
  for (let j=s.length-1; j>=0; j--) {
    if (s[j] == '(') {
      leftBrackets++;
    } else {
      rightBrackets++;
    }
    
    if (leftBrackets === rightBrackets) max = Math.max(max, 2*leftBrackets);
    
    if (rightBrackets < leftBrackets) {
      leftBrackets = 0;
      rightBrackets = 0;
    }
  }
  
  return max;
};

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        left_brackets = 0
        right_brackets = 0
        longest = 0
        
        # Check from left to right
        for i in range(len(s)):
          if s[i] == '(':
            left_brackets += 1
          else:
            right_brackets += 1
            
          if left_brackets == right_brackets:
            if right_brackets * 2 > longest:
              longest = right_brackets * 2
          
          if right_brackets > left_brackets:
            left_brackets = 0
            right_brackets = 0
        
        
        # Check from right to left
        left_brackets = 0
        right_brackets = 0
        for i in reversed(range(len(s))):
          if s[i] == '(':
            left_brackets += 1
          else:
            right_brackets += 1
            
          if left_brackets == right_brackets:
            if left_brackets * 2 > longest:
              longest = left_brackets * 2
          
          if left_brackets > right_brackets:
            left_brackets = 0
            right_brackets = 0
            
        return longest
            